Catch more circular dependencies

Dr15Jones opened the pull request:
The HLT had a case where a circular dependency was missed because the modules involved were on multiple Paths. This change now catches that case and includes an addition unit test.

Dr15Jones committed:
Catch more circular dependencies

The HLT had a case where a circular dependency was missed because the modules involved were on multiple Paths. This change now catches that case and includes an addition unit test.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20502/678

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

FWCore/Framework


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

Tested at: 30473755e9abc653874f1228b3dfc563d1490d09


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8dd9a2c51f009a80e5374a9b450f6fa76f0f5189
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22955/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22955/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22955/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v31 --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --fileout file:RelVal_Raw_2e34v31_MC.root : FAILED - time: date Wed Sep 13 19:24:23 2017-date Wed Sep 13 19:22:40 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v31.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:24:23 2017-date Wed Sep 13 19:22:40 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:2e34v31,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v31 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v31_MC.root --fileout file:RelVal_Raw_2e34v31_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:24:23 2017-date Wed Sep 13 19:22:40 2017 s - exit: 21504
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v30 --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --fileout file:RelVal_Raw_2e34v30_MC.root : FAILED - time: date Wed Sep 13 19:24:23 2017-date Wed Sep 13 19:22:44 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v30.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:24:23 2017-date Wed Sep 13 19:22:44 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:2e34v30,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v30 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v30_MC.root --fileout file:RelVal_Raw_2e34v30_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:24:23 2017-date Wed Sep 13 19:22:44 2017 s - exit: 21504
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_GRun --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --fileout file:RelVal_Raw_GRun_MC.root : FAILED - time: date Wed Sep 13 19:24:49 2017-date Wed Sep 13 19:23:05 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:24:49 2017-date Wed Sep 13 19:23:05 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_GRun_MC.root --fileout file:RelVal_Raw_GRun_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:24:49 2017-date Wed Sep 13 19:23:05 2017 s - exit: 21504
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_25ns : FAILED - time: date Wed Sep 13 19:25:08 2017-date Wed Sep 13 19:23:19 2017 s - exit: 20736
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 : FAILED - time: date Wed Sep 13 19:26:38 2017-date Wed Sep 13 19:24:32 2017 s - exit: 20736
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=HeavyIons -n 10 --conditions auto:run2_mc_HIon --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016,Run2_HI --fileout file:RelVal_Raw_HIon_MC.root : FAILED - time: date Wed Sep 13 19:26:22 2017-date Wed Sep 13 19:24:48 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:26:22 2017-date Wed Sep 13 19:24:48 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --mc --scenario=HeavyIons -n 10 --conditions auto:run2_mc_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016,Run2_HI --processName=HLTRECO --filein file:RelVal_Raw_HIon_MC.root --fileout file:RelVal_Raw_HIon_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:26:22 2017-date Wed Sep 13 19:24:48 2017 s - exit: 21504
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_PIon --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --fileout file:RelVal_Raw_PIon_MC.root : FAILED - time: date Wed Sep 13 19:26:53 2017-date Wed Sep 13 19:25:08 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:26:53 2017-date Wed Sep 13 19:25:08 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_PIon_MC.root --fileout file:RelVal_Raw_PIon_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:26:53 2017-date Wed Sep 13 19:25:08 2017 s - exit: 21504
cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Wed Sep 13 19:28:26 2017-date Wed Sep 13 19:26:23 2017 s - exit: 20736
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v22 --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --fileout file:RelVal_Raw_2e34v22_MC.root : FAILED - time: date Wed Sep 13 19:28:59 2017-date Wed Sep 13 19:26:47 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_2e34v22.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:28:59 2017-date Wed Sep 13 19:26:47 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:2e34v22,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_2e34v22 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_2e34v22_MC.root --fileout file:RelVal_Raw_2e34v22_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:28:59 2017-date Wed Sep 13 19:26:47 2017 s - exit: 21504
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake1 --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_25ns --fileout file:RelVal_Raw_Fake1_MC.root : FAILED - time: date Wed Sep 13 19:28:27 2017-date Wed Sep 13 19:27:00 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake1.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:28:27 2017-date Wed Sep 13 19:27:00 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:Fake1,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake1 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_25ns --processName=HLTRECO --filein file:RelVal_Raw_Fake1_MC.root --fileout file:RelVal_Raw_Fake1_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:28:27 2017-date Wed Sep 13 19:27:00 2017 s - exit: 21504
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_PRef --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --fileout file:RelVal_Raw_PRef_MC.root : FAILED - time: date Wed Sep 13 19:30:17 2017-date Wed Sep 13 19:28:26 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:30:17 2017-date Wed Sep 13 19:28:26 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --processName=HLTRECO --filein file:RelVal_Raw_PRef_MC.root --fileout file:RelVal_Raw_PRef_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:30:17 2017-date Wed Sep 13 19:28:26 2017 s - exit: 21504
cmsDriver.py TTbar_Tauola_8TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW  --mc --scenario=pp -n 10 --conditions auto:run1_mc_Fake --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --fileout file:RelVal_Raw_Fake_MC.root : FAILED - time: date Wed Sep 13 19:29:30 2017-date Wed Sep 13 19:28:28 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:29:30 2017-date Wed Sep 13 19:28:28 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:Fake,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run1_mc_Fake --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --processName=HLTRECO --filein file:RelVal_Raw_Fake_MC.root --fileout file:RelVal_Raw_Fake_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:29:30 2017-date Wed Sep 13 19:28:28 2017 s - exit: 21504
cmsDriver.py TTbar_Tauola_13TeV_TuneCUETP8M1_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake2 --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --fileout file:RelVal_Raw_Fake2_MC.root : FAILED - time: date Wed Sep 13 19:30:45 2017-date Wed Sep 13 19:29:03 2017 s - exit: 20736
cmsRun /cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_9_4_X_2017-09-13-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake2.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Sep 13 19:30:45 2017-date Wed Sep 13 19:29:03 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:Fake2,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake2 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --processName=HLTRECO --filein file:RelVal_Raw_Fake2_MC.root --fileout file:RelVal_Raw_Fake2_MC_HLT_RECO.root : FAILED - time: date Wed Sep 13 19:30:45 2017-date Wed Sep 13 19:29:03 2017 s - exit: 21504



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8dd9a2c51f009a80e5374a9b450f6fa76f0f5189
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22955/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22955/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Dr15Jones committed:
Properly handle circular dependency finding for modules on multiple Paths

Modules on multiple paths have to be handled carefully since the paths the circular dependency is scan are initially ambiguous.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20502/702

cmsbuild commented:
Pull request #20502 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
@davidlange6 Sorry, should have put this on hold. I'm testing all the RelVals locally. I already know failures will happen with the base-line since I haven't commited to this pull request yet a fix for a newly discovered circular dependency. See #20504

cmsbuild commented:
-1

Tested at: 360f49086cf82c5172a5a4f0a0fd996c61ea9a84


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/3f63597e6b16413d5593177262364795df467beb
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22982/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22982/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22982/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/3f63597e6b16413d5593177262364795df467beb
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22982/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22982/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Dr15Jones committed:
Fix circular dependency in validation_step path

The sequence bdHadronTrackValidationSeq was causing a circular module
dependency in the validation_step path. This was not noticed originally
because the modules causing the circular dependency were being run
in another path. Changing bdHadronTrackValidationSeq to use a Task
solved the problem. This did not make a substantial change to the
runtime order of modules because mucn of that sequence's modules
were already on other Tasks.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20502/704

cmsbuild commented:
Pull request #20502 was updated. @smuzaffar, @Dr15Jones, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @vanbesien can you please check and sign again.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
The test failure in workflow 25.0 is new.

Dr15Jones commented:
I have comfirmed that workflow 25.0 does have a circular dependency.

Dr15Jones committed:
Fix circular dependency in pathALCARECOEcalUncal*Electron paths

An improvement to the module circular dependency finder uncovered
a circular dependency in all pathALCARECOEcalUncal*Electron. The
Fix was to move the problematic modules to a Task.

The underlying problem was
-ecalDigis was after WenuSelector on the ALCA paths
-WenuSelector consumes pfMet
-pfMet is after particleFlowRecHitECAL on path reconstruction_step
-particleFlowRecHitECAL consumes ecalDigis
This only happened to work because ecalDigis is also on the path
raw2digi_step. However, if there was an exception on raw2digi_step,
ecalDigis would not be run from that path and then the job would
deadlock because of the remaining circular dependency.

cmsbuild commented:
The code-checks are being triggered in jenkins.

davidlange6 commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20502/707

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #20502 was updated. @ghellwig, @smuzaffar, @Dr15Jones, @vazzolini, @kmaeshima, @arunhep, @cerminar, @dmitrijus, @cmsbuild, @franzoni, @vanbesien, @lpernie can you please check and sign again.

cmsbuild commented:
+1
Tested at: 8fec79ea4c53489e4dce02e75d839007185ffa46
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22985/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/3f63597e6b16413d5593177262364795df467beb
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22985/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22985/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/22985/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2646869
 - DQMHistoTests: Total failures: 204
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2646476
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

ghellwig commented on Calibration/EcalAlCaRecoProducers/python/ALCARECOEcalUncalIsolElectron_cff.py:
```diff
@@ -26,7 +26,7 @@
 ecalUncalibRecHitSequence53X = cms.Sequence(ecalGlobalUncalibRecHit * ecalDetIdToBeRecovered)
 
-uncalibRecHitSeq = cms.Sequence( (ecalDigis + ecalPreshowerDigis) * ecalUncalibRecHitSequence)
+uncalibRecHitSeq = cms.Sequence(ecalUncalibRecHitSequence, cms.Task(ecalDigis, ecalPreshowerDigis))
```
@Dr15Jones is there some documentation to understand the reasoning behind this change?

Dr15Jones replied:
Is the commit message sufficient?
https://github.com/cms-sw/cmssw/pull/20502/commits/8fec79ea4c53489e4dce02e75d839007185ffa46

ghellwig replied:
I guess my problem is that I am a bit lacking understanding of the `cms.Task` concept.
What extra information does it provide to wrap `ecalDigis + ecalPreshowerDigis` in a `cms.Task`.
Sorry for my previous ignorance...

Dr15Jones replied:
You can find the full documentation about Tasks here
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideAboutPythonConfigFile#Task_Objects

Essentially, modules on a Task will be run when their data products are first needed by another module (it is what we used to call 'unscheduled'). In contrast, a Sequence runs its modules in a specific order. However, in this case, the Sequence order was causing an circular dependency between the modules. Moving the modules to a Task, however, allows the framework to run the modules only when needed and this avoids the circle (since the order the modules run is strictly based on data dependency and not by an external ordering).

ghellwig replied:
After making a sketch of your commit message it was literally a circular dependency.
Thanks for the thorough explanation!

ghellwig commented:
+1

Dr15Jones commented:
+1

Dr15Jones commented:
I ran the full runTheMatrix.py and addOnTests.py suites with this change and no additional circular dependencies were found.

Dr15Jones commented:
@dmitrijus @vanbesien @kmaeshima @vazzolini Please review and then sign for dqm

davidlange6 commented on Validation/RecoB/python/BDHadronTrackValidation_cff.py:
```diff
@@ -13,15 +13,15 @@
-                                        * BDHadronTrackMonitoringAnalyze
+bdHadronTrackValidationSeq = cms.Sequence(BDHadronTrackMonitoringAnalyze,
+                                          cms.Task(patJetCorrectionsTask
+                                                   , patJetCharge
```
hi @Dr15Jones - I'm wondering if the real fix is to remove all of these pat* modules from the sequence (in line with them running inside the PAT sequence already)

Dr15Jones replied:
That I can't answer. I was trying for the minimum changes possible. From the framework end, having a Path depend on products from another Path is good if we want to be able to be able to be able to break those Paths into different Processes at some time in the future. Else, the framework would much prefer a Path be fully self contained where all EDProducers are on Tasks and not Sequences.

davidlange6 replied:
hi @Dr15Jones - the idea of self contained paths would be a migration.. everything is built up from the opposite approach (sequential paths that rely on each other),

unfortunately I see that this module is monitoring a module that no one has asked to run in miniAOD (patJetsBDHadron at least). This means that potentially we change miniAOD results with or without the validation sequence...I'm not sure I know the motivation for that yet.

@slava77 @arizzi

Dr15Jones replied:
The problem we have here is the existing code is an amalgam of both 'semi-self contained' and 'sequences depending on other sequences'. There is one sequence which wants some earlier sequence to define what it reads while a later sequence in the same path tries to be 'self contained' and has the modules needed by the earlier sequence.

Now keep in mind my change doesn't do anything to the existing job results since all these `pat*` modules exist on a `Path` which is guaranteed to run before the `EndPath`.

Dr15Jones replied:
@davidlange6 on the miniAOD side, are you worried that some existing module used in the miniAOD asks for one of these `pat*` and right now since it isn't there they silently fail but with this change they would now succeed?

davidlange6 replied:
Right - that or a missing/added data product in the output.

dmitrijus commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

Dr15Jones commented:
@davidlange6 when will this be merged?

Dr15Jones commented:
@davidlange6 anything I can do to help get this merged?

davidlange6 committed:
Update BDHadronTrackValidation_cff.py

remove all the miniaod components that are already running in the miniaod sequence

davidlange6 commented:
please test

davidlange6 commented:
@Dr15Jones - does this still maintain the original fix of the circular dependency?

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
@davidlange6 I'm not sure if it is circular or not. One way to find out is to run workflow 10.0, since #20504 specifically mentions that one as seeing the circular dependency once this fix was added.

Dr15Jones commented:
please test workflow 10.0

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #20502 was updated. @ghellwig, @smuzaffar, @Dr15Jones, @vazzolini, @kmaeshima, @arunhep, @cerminar, @dmitrijus, @cmsbuild, @franzoni, @vanbesien, @lpernie can you please check and sign again.

Dr15Jones commented:
@davidlange6 thinking more about your change and yes, it will also preserve the circular dependency fix. For the workflows we use containing this Sequence, moving the modules from the Sequence to an associated Task basically means those modules are ignored for scheduling purposes because those same models appear on another Path. 
So the only difference between your change and mine is what happens when someone uses that Sequence in non production cases and they have not included those pat modules in their Paths.

cmsbuild commented:
-1

Tested at: a12c8a8d016cd60686069b0496eb4b88a6d389da


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/829d9e6ab8f6d2ba0a6b15fe515c39963d9ed646
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23339/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23339/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23339/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc_l1stage1 --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_25ns : FAILED - time: date Sat Sep 30 20:55:56 2017-date Sat Sep 30 20:48:33 2017 s - exit: 17920
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 : FAILED - time: date Sat Sep 30 20:59:49 2017-date Sat Sep 30 20:54:08 2017 s - exit: 17920
cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Sat Sep 30 21:08:41 2017-date Sat Sep 30 21:02:20 2017 s - exit: 17920



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/829d9e6ab8f6d2ba0a6b15fe515c39963d9ed646
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23339/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23339/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Dr15Jones commented:
@davidlange6 looks like your change broke those workflows. Specifically patJetsBDHadron can not find patJetFlavourAssociation.

Dr15Jones commented:
@davidlange6 I looked at the summary report for the standard 5.1 workflow and `patJetFlavourAssociation` is not on a Path or EndPath, it comes from a Task.

https://github.com/cms-sw/cmssw/blob/c61529f278fb0f140a9002739fb237d8be4f183b/PhysicsTools/PatAlgos/python/mcMatchLayer0/jetFlavourId_cff.py#L43

In the code in the release, it comes in on the sequence `patJetFlavourId` which just holds the Task `patJetFlavourIdTask` that in turn contains the module.

From the summary information I also see that `patJetFlavourIdLegacy (and its Task equivalent) is not needed for 5.1 because the two modules it brings in are never run.

```
TrigReport         10         10         10          0          0 patJetFlavourAssociation
TrigReport          0          0          0          0          0 patJetFlavourAssociationLegacy
TrigReport          0          0          0          0          0 patJetPartonAssociationLegacy
TrigReport         40         10         10          0          0 patJetPartons
TrigReport          0          0          0          0          0 patJetPartonsLegacy
```
and, of course, the stuff on the EndPath always ran

```
TrigReport     0  132         10         10          0          0 patJetCharge
TrigReport     0  133         10         10          0          0 patJetPartonMatch
TrigReport     0  134         10         10          0          0 patJetGenJetMatch
TrigReport     0  135         10         10          0          0 patJetsBDHadron
TrigReport     0  136         10         10          0          0 tpClusterProducer
TrigReport     0  137         10         10          0          0 BDHadronTrackMonitoringAnalyze
```

Dr15Jones commented:
We could try adding `patJetFlavourIdTask` into your change, but it is not clear if that will just cause `patJetFlavourAssociation` to fail because it can't find data it needs.

Dr15Jones commented:
@davidlange6  The minimum change to 5.1 needed to make it run was
```diff
bdHadronTrackValidationSeq = cms.Sequence(BDHadronTrackMonitoringAnalyze,
-                                          cms.Task(patJetCorrectionsTask, patJetsBDHadron)
+                                          cms.Task(patJetCorrectionsTask, patJetCharge,tpClusterProducer,patJetFlavourIdTask,patJetPartonMatch,patJetGenJetMatch,patJetsBDHadron)
 ) 
```

davidlange6 commented:
ah - wonderful - indeed, I always enjoy things on end paths for no reason.. I guess I should just undo my change and cover up the bigger issue.

davidlange6 committed:
Update BDHadronTrackValidation_cff.py

revert my earlier change

cmsbuild commented:
The code-checks are being triggered in jenkins.

davidlange6 commented:
please test

Dr15Jones commented:
@davidlange6 thanks

Dr15Jones commented:
+1

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20502/1089

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #20502 was updated. @ghellwig, @vazzolini, @kmaeshima, @arunhep, @cerminar, @dmitrijus, @cmsbuild, @franzoni, @vanbesien, @lpernie can you please check and sign again.

cmsbuild commented:
+1
Tested at: d816ffa894e18d08f85754bfba92d205037c86bc
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23365/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/6fa527f6076a91c47418a69c11e63174d88c6c94
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23365/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23365/git-merge-result

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
+1

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20502/23365/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2760941
 - DQMHistoTests: Total failures: 245
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2760509
 - DQMHistoTests: Total skipped: 187
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

davidlange6 commented:
merge
